concordium-consensus/src/Concordium/Skov/Update.hs

```diff
@@ -161,37 +159,44 @@ processFinalization newFinBlock finRec@FinalizationRecord{..} = do
         -- to mark as dead. When we exhaust the branches we then mark blocks as finalized
         -- by increasing height.
-        -- Instead of marking blocks dead immediately we accumulate them
-        -- and a return a list. The reason for doing this is that we never
-        -- have to look up a parent block that is already marked dead.
-        let
-            pruneTrunk :: [BlockPointerType m] -- ^List of blocks to remove.
-                       -> BlockPointerType m -- ^The block that was finalized.
+        -- 
+        -- Instead of marking blocks dead immediately, we accumulate them in a list by decreasing
+        -- height. The reason for doing this is that we never have to look up a parent block that is
+        -- already marked dead.
+        let pruneTrunk :: [BlockPointerType m] -- ^List of blocks to remove.
+                       -> [BlockPointerType m] -- ^List of finalized blocks.
                        -> Branches m
-                       -> m [BlockPointerType m]
-                       -- ^ The return value is a list of blocks to mark dead, ordered
-                       -- by increasing height.
-            pruneTrunk toRemove _ Seq.Empty = return toRemove
-            pruneTrunk toRemove keeper (brs Seq.:|> l) = do
-                parent <- bpParent keeper
-                let toRemove1 = filter (/= keeper) l ++ toRemove
-                toRemove2 <- pruneTrunk toRemove1 parent brs
-                -- mark blocks as finalized now, so that blocks are marked finalized by increasing height
-                markFinalized (getHash keeper) finRec
-                logEvent Skov LLDebug $ "Block " ++ show keeper ++ " marked finalized"
-                -- Finalize the transactions of the surviving block.
-                -- (This is handled in order of finalization.)
-                finalizeTransactions (getHash keeper) (blockSlot keeper) (blockTransactions keeper)
-                ati <- bpTransactionAffectSummaries keeper
-                bcTime <- getSlotTimestamp (blockSlot keeper)
-                let ctx = BlockContext{
-                        bcHash = getHash keeper,
-                        bcHeight = bpHeight keeper,
-                        ..}
-                flushBlockSummaries ctx ati =<< getSpecialOutcomes =<< blockState keeper
-                return toRemove2
-
-        toRemoveFromTrunk <- pruneTrunk [] newFinBlock (Seq.take pruneHeight oldBranches)
+                       -> m ([BlockPointerType m], [BlockPointerType m])
+                       -- ^ The return value is a list of blocks to mark dead and a list of blocks
+                       -- to mark finalized, ordered by increasing height.
+            pruneTrunk toRemove toFinalize Seq.Empty = return (toRemove, toFinalize)
+            pruneTrunk toRemove toFinalize (brs Seq.:|> l) = do
+              let keeper = head toFinalize
+                  toRemove1 = toRemove ++ filter (/= keeper) l
```

Comment:
Since we are traversing the branches by decreasing height, I believe this should be as it was, so
```
filter (/= keeper) l ++ toRemove
```
since that will add blocks with lower height at the beginning of the list.